### PR TITLE
Scvi project

### PR DESCRIPTION
**Pull request by yoavnahum21**
-   [ ] Closes #xxxx (Replace xxxx with the GitHub issue number)
-   [ ] Tests added and passed if fixing a bug or adding a new feature
-   [ ] All code checks passed.
-   [ ] Added type annotations to new arguments/methods/functions.
-   [ ] Added an entry in the latest `docs/release_notes/index.md` file if fixing a bug or adding a new feature.
-   [ ] If the changes are patches for a version, I have added the `on-merge: backport to x.x.x` label.

### CODE REVIEW

File: scvi/model/_scvi.py [hunk 1]
```diff
@@ -127,7 +128,10 @@ def __init__(
             library_log_means, library_log_vars = _init_library_size(
                 self.adata_manager, n_batch
             )
-
+        [self.batch_0, self.batch_1, self.batch_2, self.batch_3, self.batch_4] = [torch.tensor(adata.X[adata.obs['_scvi_batch'] == 0]), torch.tensor(adata.X[adata.obs['_scvi_batch'] == 1]),
+                                                                                  torch.tensor(adata.X[adata.obs['_scvi_batch'] == 2]), torch.tensor(adata.X[adata.obs['_scvi_batch'] == 3]),
+                                                                                  torch.tensor(adata.X[adata.obs['_scvi_batch'] == 4])]
+        self.batches = [self.batch_0, self.batch_1, self.batch_2, self.batch_3, self.batch_4]
```
Comment (martinkim0) [inline scvi/model/_scvi.py]: Look into this tutorial for data handling strategy: https://docs.scvi-tools.org/en/stable/tutorials/notebooks/dev/data_tutorial.html

Comment (martinkim0) [reply scvi/model/_scvi.py]: Think of edge cases: what if there are more than 5 batches?

File: scvi/module/_vae.py [hunk 2]
```diff
@@ -100,6 +101,8 @@ def __init__(
         n_input: int,
         n_batch: int = 0,
         n_labels: int = 0,
+        mode: str = "fast",
+        batches: torch.tensor = None,
```
Comment (martinkim0) [inline scvi/module/_vae.py:105]:
```suggestion
        batches: torch.tensor = None,
        beta: Tunable[float] = 1.0,
```

File: scvi/module/_vae.py [hunk 3]
```diff
@@ -131,10 +134,15 @@ def __init__(
         self.gene_likelihood = gene_likelihood
         # Automatically deactivate if useless
         self.n_batch = n_batch
+        self.batches = batches
         self.n_labels = n_labels
+        self.beta = torch.tensor(1.0)
```
Comment (martinkim0) [inline scvi/module/_vae.py]:
```suggestion
        self.beta = beta
```

File: scvi/module/_vae.py [hunk 4]
```diff
@@ -490,7 +554,6 @@ def loss(
         return LossOutput(
             loss=loss, reconstruction_loss=reconst_loss, kl_local=kl_local
```
Comment (martinkim0) [inline scvi/module/_vae.py]: Add MMD loss to LossOutput so that it is available in `model.history`

File: scvi/module/_vae.py [hunk 5]
```diff
@@ -131,10 +134,15 @@ def __init__(
         self.gene_likelihood = gene_likelihood
         # Automatically deactivate if useless
         self.n_batch = n_batch
+        self.batches = batches
         self.n_labels = n_labels
+        self.beta = torch.tensor(1.0)
+        self.mode = mode
+        self.epoch = 0
+        self.error = []
+        self.mmdloss = self._compute_mmd_loss(self.batches, self.mode)
```
Comment (martinkim0) [inline scvi/module/_vae.py]: Not necessary to compute MMD loss here, just within `loss` is fine

File: scvi/module/_vae.py (hunk 4 above)
Comment (martinkim0) [reply scvi/module/_vae.py]: Might need to go through `LossOutput` code in `scvi.module.base._base_module` and add changes there too.

**Review by martinkim0 — COMMENTED** [review]
Good progress!
